pkgs/development/python-modules/pipx/default.nix

```diff
@@ -0,0 +1,64 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, userpath
+, argcomplete
+, packaging
+, pytest
+}:
+
+buildPythonPackage rec {
+  pname = "pipx";
+  version = "0.15.5.0";
+
+  # no tests in the pypi tarball, so we directly fetch from github
+  src = fetchFromGitHub {
+    owner = "pipxproject";
+    repo = pname;
+    rev = version;
+    sha256 = "13z032i8r9f6d09hssvyjpxjacb4wgms5bh2i37da2ili9bh72m6";
+  };
+
+  propagatedBuildInputs = [ userpath argcomplete packaging ];
+
+  checkInputs = [ pytest ];
+
+  preCheck = ''
+    export HOME=$(mktemp -d)
+    python -m venv .
+    source ./bin/activate
+  '';
+
+  # disable tests, which require internet connection
+  disabledTests = [
+    "test_install"
+    "test_inject"
+    "test_missing_interpreter"
+    "test_package_install"
+    "test_package_inject"
+    "test_reinstall_all"
+    "test_cache"
+    "test_run_script_from_internet"
+    "test_run_ensure_null_pythonpath"
+    "test_runpip"
+    "test_uninstall"
+    "test_uninstall_all"
+    "test_upgrade"
+    "test_upgrade_all"
+  ];
+
+  disabledTestsExpression = lib.concatMapStringsSep " and " (s: "not " + s) disabledTests;
+
+  checkPhase = ''
+    runHook preCheck
+    python -m pytest -k "${disabledTestsExpression}"
+  '';
```

Comment:
you can use pytestCheckHook instead

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pytestcheckhook

Reply:
It seems like `preCheck` hook is not executed after i removed these lines and added `pytestCheckHook` to `checkInputs` (Tests require venv issue). Have no idea, why

Reply:
running the pre and post check hooks only really matters if you expect the derivation to be overriden, and for people to re-use the existing checkPhase, but just add a pre/postCheck.

Since the package knows how to check itself, the usefulness of those hooks are significantly less.

And if you're looking for completeness, then you'll want to add `runHook postCheck` as well.

Reply:
Ok, but i should instantiate and activate venv anyway, because tests require it. Actually everything should work after substituting these lines with `pytestCheckHook` as it runs `preCheck` hook during its `checkPhase` https://github.com/NixOS/nixpkgs/blob/29062cec8d214013a8e1f944d16e32c36294e0d0/pkgs/development/interpreters/python/hooks/pytest-check-hook.sh#L29-L31 but something is wrong and hook is not executed

Reply:
for the preCheck, it's just creating a venv to make some assertion valid as part of the test suite. However, the underlying assertion is just making sure "venv" is present, which is in all python interpreters > 3.3.

I'm able to run all the tests with just:
```
{ lib
, buildPythonPackage
, fetchFromGitHub
, pythonOlder
, pytestCheckHook
, userpath
, argcomplete
, packaging
, pytest
}:

buildPythonPackage rec {
  pname = "pipx";
  version = "0.15.5.0";
  disabled = pythonOlder "3.6";

  # no tests in the pypi tarball, so we directly fetch from github
  src = fetchFromGitHub {
    owner = "pipxproject";
    repo = pname;
    rev = version;
    sha256 = "13z032i8r9f6d09hssvyjpxjacb4wgms5bh2i37da2ili9bh72m6";
  };

  propagatedBuildInputs = [ userpath argcomplete packaging ];

  postPatch = ''
    substituteInPlace tests/helpers.py \
      --replace 'assert getattr(sys, "base_prefix", sys.prefix) != sys.prefix, "Tests require venv"' ""
  '';

  checkInputs = [ pytestCheckHook ];

  preCheck = ''
    export HOME=$(mktemp -d)
  '';

  # disable tests, which require internet connection
  disabledTests = [
    "install"
    "inject"
    "ensure_null_pythonpath"
    "missing_interpreter"
    "cache"
    "internet"
    "runpip"
    "upgrade"
  ];

  meta = with lib; {
    description =
      "Install and Run Python Applications in Isolated Environments";
    homepage = "https://github.com/pipxproject/pipx";
    license = licenses.mit;
    maintainers = with maintainers; [ yevhenshymotiuk ];
  };
}
```

Reply:
I made a PR https://github.com/pipxproject/pipx/pull/476 to fix this properly

Reply:
Should I commit this `postPatch` for now?

Reply:
actually, closed 476 in favor of https://github.com/pipxproject/pipx/pull/477

Reply:
> Should I commit this `postPatch` for now?

I think so. Or we could way to see if upstream wants to do a patch release, although i didn't really touch their production code.